rewrite-core/src/main/java/org/openrewrite/RecipeScheduler.java

```diff
@@ -60,57 +57,39 @@ private LargeSourceSet runRecipeCycles(Recipe recipe, LargeSourceSet sourceSet,
+            ctxWithWatch.putCycle(cycle);
+            after.beforeCycle();
+
+            // pre-transformation scanning phase where there can only be modifications to capture exceptions
+            // occurring during the scanning phase
+            if (hasScanningRecipe(recipe)) {
+                after = cycle.scanSources(after);
             }
 
-            // this root cursor is shared by all `TreeVisitor` instances used created from `getVisitor` and
-            // single source applicable tests so that data can be shared at the root (especially for caching
-            // use cases like sharing a `JavaTypeCache` between `JavaTemplate` parsers).
-            Cursor rootCursor = new Cursor(null, Cursor.ROOT_VALUE);
-            try {
-                RecipeRunCycle<LargeSourceSet> cycle = new RecipeRunCycle<>(recipe, i, rootCursor, ctxWithWatch,
-                        recipeRunStats, sourceFileResults, errorsTable, LargeSourceSet::edit);
-                ctxWithWatch.putCycle(cycle);
-                after.beforeCycle();
-
-                // pre-transformation scanning phase where there can only be modifications to capture exceptions
-                // occurring during the scanning phase
-                if (hasScanningRecipe(recipe)) {
-                    after = cycle.scanSources(after);
-                }
-
-                // transformation phases
-                after = cycle.generateSources(after);
-                after = cycle.editSources(after);
-
-                boolean anyRecipeCausingAnotherCycle = false;
-                for (Recipe madeChanges : cycle.getMadeChangesInThisCycle()) {
-                    if (madeChanges.causesAnotherCycle()) {
-                        anyRecipeCausingAnotherCycle = true;
-                    }
-                }
-
-                if (i >= minCycles &&
-                    (cycle.getMadeChangesInThisCycle().isEmpty() || !anyRecipeCausingAnotherCycle)) {
-                    after.afterCycle(true);
-                    break;
-                }
+            // transformation phases
+            after = cycle.generateSources(after);
+            after = cycle.editSources(after);
 
-                after.afterCycle(i == maxCycles);
-                ctxWithWatch.resetHasNewMessages();
-            } finally {
-                // Clear any messages that were added to the root cursor during the cycle. This is important
-                // to avoid leaking memory in the case when a recipe defines a static TreeVisitor. That
-                // TreeVisitor will still contain a reference to this rootCursor and any messages in it
-                // after recipe execution completes. The pattern of holding a static TreeVisitor isn't
-                // recommended, but isn't possible for us to guard against at an API level, and so we are
-                // defensive about memory consumption here.
-                rootCursor.clearMessages();
-            }
+            after.afterCycle(true);
```

Comment:
If we proceed with this, we should probably at least also rename this method. We probably still need some kind of life cycle method, so I don't think we can remove it.